GH-40783: [C++] Re-order loads and stores in MemoryPoolStats update

felipecrv opened the pull request:
### Rationale for this change

Issue loads as soon as possible so the latency of waiting for memory is masked by doing other operations.

### What changes are included in this PR?

 - Make all the read-modify-write operations use `memory_order_acq_rel`
 - Make all the loads and stores use `memory_order_acquire`/`release` respectively
 - Statically specialize the implementation of `UpdateAllocatedBytes` so `bytes_allocated_` can be updated without waiting for the load of the old value

### Are these changes tested?

By existing tests.

* GitHub Issue: #40783

amoeba commented:
@ursabot please benchmark

mapleFU reviewed (commented):
I like this change, but personally, I think `relaxed` is enough here( relaxed gurantee "atomic", acq-rel gurantees "happens-before") ?

acq-rel might be used in the case below:

```
Global Ctx t
std::atomic<bool> condition{false};

void Acq() {
  set Ctx
  condition.store(true, std::memory_order_acq);
}

void Rel() {
  while (!test condition ) { yield(); }
  read Ctx
}
```

For MemoryPool, do we really need a acq-rel here?

mapleFU commented on cpp/src/arrow/memory_pool.h:
```diff
@@ -36,43 +37,58 @@ namespace internal {
+  //
+  // max_memory_, total_allocated_bytes_, and num_allocs_ only go up (they are
+  // monotonically increasing) which can allow some optimizations.
+  std::atomic<int64_t> bytes_allocated_{0};
```
Should it aligned by cache-line for performance ( https://en.cppreference.com/w/cpp/thread/hardware_destructive_interference_size )?

felipecrv replied:
I did put all the atomics on the same cache line, but I used `64` instead of this STL thing that will grow like the inflation.

mapleFU commented on cpp/src/arrow/memory_pool.h:
```diff
@@ -36,43 +37,58 @@ namespace internal {
 
-  int64_t num_allocations() const { return num_allocs_.load(); }
+ public:
+  int64_t max_memory() const { return max_memory_.load(std::memory_order_acquire); }
```
Would relaxed ok here?

pitrou commented:
Does it actually change anything for x86?

pitrou commented:
If you're really interested in reducing the contention costs for MemoryPool statistics, then I would suggest taking a look at https://travisdowns.github.io/blog/2020/07/06/concurrency-costs.html

felipecrv commented:
> Does it actually change anything for x86?

Nothing, unless the compiler decides to re-order the loads and the stores. Which it didn't in this case.

I suspect my change regarding `max_memory_` didn't lead to a cheaper sequence of instructions — I assumed `max_memory_.store(allocated, seq_cst)` (written as `max_memory_ = allocated;` in the code) was `lock`-prefixed and more expensive (combined with the `seq_cst` load) than the `compare_exchange` I added.


> If you're really interested in reducing the contention costs for MemoryPool statistics, then I would suggest taking a look at https://travisdowns.github.io/blog/2020/07/06/concurrency-costs.html

Not just the contention (the least of the issues really), but the time it takes to perform all the memory operations.

mapleFU commented:
> Does it actually change anything for x86?

Wouldn't acq-rel cheaper than original total-ordering?

felipecrv commented:
> > Does it actually change anything for x86?
> 
> Wouldn't acq-rel cheaper than original total-ordering?

Only on architectures with a weaker memory model (e.g. ARM). x86 guarantees that all the stores are ordered (not immediately visible, but they are ordered). [1] explains this much better than I ever could.

`acq-rel` can enable the *compiler* to re-order operations if it decides that can unlock optimizations. Not the case here, so I'm manually experimenting with different orderings to mask the latency of fetch/load operations on atomic variables.

[1] https://research.swtch.com/hwmm#x86

felipecrv commented:
I pushed some re-ordering of loads and stores that I believe can work better on CPUs with higher latency on the memory system [1]. Note that my code updates `max_memory_` correctly (I removed the comment about "no need to be rigorous"). The `lock cmpxchg` that I introduced never appears in profiles, so I'm keeping it. I also suspect it would be beneficial on contented workloads because we can give up on updating `max_memory_` if another threads increases it before the current thread.

When benchmarking on an Intel CPU, compared to the baseline, this version doesn't cost less, but it should improve the situation on CPUs that are spending more time [1] in the load issued by `bytes_allocated_.fetch_add(diff)` (`lock xadd` instruction on x86). My hypothesis is that by not immediately using the result of the `xadd`, the CPU can wait for the memory system in the background while performing other operations. It also clusters all the `lock`-prefixed instructions together which is recommended practice.

Annotated `perf report` of `arrow::memory_pool::internal::JemallocAllocator::AllocateAligned` after these changes:

```asm
#   Percent│       push %rbp
           │       push %r15
           │       push %r14
           │       push %r13
           │       push %r12
           │       push %rbx
           │       push %rax
           │       mov  %r8,%r13
           │       mov  %rcx,%rbx
           │       mov  %rdx,%r12
           │       mov  %rsi,%r15
           │       mov  %rdi,%r14
           │       incb 0x13f6583(%rip)        # 61bc853 <__TMC_END__+0x13935b>
1     5.31 │       xor  %edi,%edi
           │       mov  %rdx,%rsi
           │     → call __sanitizer_cov_trace_const_cmp8@plt
           │       test %r12,%r12
           │     ↓ js   6c
1     5.56 │       incb 0x13f6570(%rip)        # 61bc855 <__TMC_END__+0x13935d>
           │       mov  %rsp,%rdi
           │       mov  %r12,%rsi
           │       mov  %rbx,%rdx
           │       mov  %r13,%rcx
           │     → call arrow::memory_pool::internal::JemallocAllocator::AllocateAligned@plt
           │       test %r14,%r14
           │     ↓ je   136
           │       incb 0x13f6552(%rip)        # 61bc857 <__TMC_END__+0x13935f>
           │       mov  (%rsp),%rax
           │       mov  %rax,(%r14)
           │       test %rax,%rax
           │     ↓ je   8b
           │       incb 0x13f6541(%rip)        # 61bc858 <__TMC_END__+0x139360>
           │     ↓ jmp  11e
           │ 6c:   incb 0x13f6532(%rip)        # 61bc854 <__TMC_END__+0x13935c>
           │       lea  typeinfo name for arrow::json::TableReader+0x22a,%rdx
           │       mov  %r14,%rdi
           │       mov  $0x4,%esi
           │     → call arrow::Status::FromArgs<char const (&) [21]>@plt
           │     ↓ jmp  11e
           │ 8b:   incb 0x13f6518(%rip)        # 61bc859 <__TMC_END__+0x139361>
1     5.54 │       mov  0x40(%r15),%rbx
           │       mov  %r12,%rsi
4    22.34 │       lock xadd %rsi,0x48(%r15)
4    22.47 │       lock add  %r12,0x50(%r15)
6    33.09 │       lock incq 0x58(%r15)
           │       mov  %rsi,%r13
           │       add  %r12,%r13
           │     ↓ jo   14a
           │       incb 0x13f64f1(%rip)        # 61bc85b <__TMC_END__+0x139363>
           │       mov  %rbx,%rdi
           │       mov  %r13,%rsi
           │     → call __sanitizer_cov_trace_cmp8@plt
           │       cmp  %r13,%rbx
1     5.69 │     ↓ jge  103
           │       incb 0x13f64dd(%rip)        # 61bc85d <__TMC_END__+0x139365>
           │ d0:   incb 0x13f64d8(%rip)        # 61bc85e <__TMC_END__+0x139366>
           │       mov  %rbx,%rax
           │       lock cmpxchg %r13,0x40(%r15)
           │       sete %bpl
           │       mov  %rax,%rbx
           │       mov  %rax,%rdi
           │       mov  %r13,%rsi
           │     → call __sanitizer_cov_trace_cmp8@plt
           │       test %bpl,%bpl
           │     ↓ jne  10b
           │       cmp  %r13,%rbx
           │     ↓ jge  10b
           │       incb 0x13f64af(%rip)        # 61bc860 <__TMC_END__+0x139368>
           │     ↑ jmp  d0
           │103:   incb 0x13f64a3(%rip)        # 61bc85c <__TMC_END__+0x139364>
           │     ↓ jmp  111
           │10b:   incb 0x13f649e(%rip)        # 61bc85f <__TMC_END__+0x139367>
           │111:   incb 0x13f649a(%rip)        # 61bc861 <__TMC_END__+0x139369>
           │       movq $0x0,(%r14)
           │11e:   incb 0x13f648e(%rip)        # 61bc862 <__TMC_END__+0x13936a>
           │       mov  %r14,%rax
           │       add  $0x8,%rsp
           │       pop  %rbx
           │       pop  %r12
           │       pop  %r13
           │       pop  %r14
           │       pop  %r15
           │       pop  %rbp
           │     ← ret
```

Benchmarks (based on the old code) on a Zen 3 CPU show that the CPU can get stuck waiting for the value produced by `lock xadd` instead of progressing:

```asm
    0.18 |    lock xadd %rax,(%rdi)
   80.73 |    add    %rsi,%rax
```

Doing useful work that doesn't depend on `%rax` (the result of `lock xadd`) should mask the latency of the memory load across the memory system.

From [1]: 

> In Zen 3, a single 32MB L3 cache pool is shared among all 8 cores in a chiplet, vs. Zen 2's two 16MB pools each shared among 4 cores in a core complex, of which there were two per chiplet. This new arrangement improves the cache hit rate as well as performance in situations that require cache data to be exchanged among cores, but increases cache latency from 39 cycles in Zen 2 to 46 clock cycles and halves per-core cache bandwidth, although both problems are partially mitigated by higher clock speeds. 

[1] https://en.wikipedia.org/wiki/Zen_3#Features

mapleFU commented:
Which benchmark should I run? I'd like to testing on my M1 Pro

mapleFU commented on cpp/src/arrow/memory_pool.h:
```diff
@@ -35,44 +36,62 @@ namespace internal {
-  int64_t bytes_allocated() const { return bytes_allocated_.load(); }
-
-  int64_t total_bytes_allocated() const { return total_allocated_bytes_.load(); }
+class alignas(/*CacheLineSize=*/64) MemoryPoolStats {
```
This is the alignas of stats, not the alignas of member?

felipecrv replied:
The members will start at a 64-byte boundary. The members have alignment that's lower than 64 — int64 are 8-byte aligned.

felipecrv replied:
Sure. Pushing soon.

mapleFU commented:
Still not understanding that. I think `xadd` is enough, is `lock xadd` neccessary? See cases above:

1. https://github.com/apache/kudu/blob/647726ad6b2aab0c6a6d34e16e027debd8a827eb/src/kudu/util/high_water_mark.h#L29
2. https://github.com/facebook/rocksdb/blob/6ddfa5f06140c8d0726b561e16dc6894138bcfa0/monitoring/histogram.cc#L76

I think just relaxed is enough here.

wgtmac commented on cpp/src/arrow/memory_pool.h:
```diff
@@ -35,44 +36,62 @@ namespace internal {
-  int64_t bytes_allocated() const { return bytes_allocated_.load(); }
-
-  int64_t total_bytes_allocated() const { return total_allocated_bytes_.load(); }
+class alignas(/*CacheLineSize=*/64) MemoryPoolStats {
```
Do you want to follow the pattern below:
```
#ifdef __cpp_lib_hardware_interference_size
    using std::hardware_constructive_interference_size;
    using std::hardware_destructive_interference_size;
#else
    // 64 bytes on x86-64 │ L1_CACHE_BYTES │ L1_CACHE_SHIFT │ __cacheline_aligned │ ...
    constexpr std::size_t hardware_constructive_interference_size = 64;
    constexpr std::size_t hardware_destructive_interference_size = 64;
#endif
```

felipecrv replied:
I prefer to fix this on 64-byte and have fixed alignment/sizes of objects (instead of relying on a platform/implementation dependent constant). If a new platform shows up where 64 is not enough and benchmarks prove that a higher alignment is necessary we can consider the size cost against the false-sharing problems and bump it up.

felipecrv commented on cpp/src/arrow/memory_pool.h:
```diff
@@ -35,44 +36,62 @@ namespace internal {
+    auto max_memory = max_memory_.load(std::memory_order_relaxed);
+    auto old_bytes_allocated =
+        bytes_allocated_.fetch_add(diff, std::memory_order_acq_rel);
+    if constexpr (IsFree) {
```
This even affected instruction selection in a productive way. When performing a `Free` we don't care about the value of `bytes_allocated_` because we know it won't grow `max_memory_`.

So when `IsFree = true` the compiler will emit `lock addq %r15, 8(%rax)` — `addq` will store the result far away in memory but we don't have to wait for the previous value. It's a fire-and-forget operation to the CPU.

When `IsFree = false` (diff > 0), the compiler will emit `lock xaddq %rdx, 8(%rcx)`. `xadd` is the Exchange Add instruction. The CPU will first load the value stored in memory and then issue the add+store. When `old_bytes_allocated` is used (by the end of this function) the value should have traveled from memory to the register. Only the add+store can be fire-and-forget (from the PoV of the CPU core), but it has to wait for the loaded value.

pitrou replied:
I'll note that the desired goal should not be to improve assembler traces in a theoretical way, but to reduce potential overhead (ideally with benchmark numbers, but I realize those might require execution on some particular machine).

felipecrv replied:
My benchmarks diff come out *neutral* so I had to use theory to explain the benchmarks I got from others. Removing data dependencies is good: I've seen concrete situations where performing a load early helped benchmarks. Unfortunately I can't prove it's helping here.

pitrou commented on cpp/src/arrow/memory_pool.h:
```diff
@@ -35,44 +36,62 @@ namespace internal {
-      max_memory_ = allocated;
-    }
+  int64_t bytes_allocated() const {
+    return bytes_allocated_.load(std::memory_order_acquire);
```
Perhaps add a private `static constexpr auto kMemoryOrder = std::memory_order_acquire` at the class level?

felipecrv replied:
(?) Different memory orders are used depending on the operation.

pitrou commented on cpp/src/arrow/memory_pool.h:
```diff
@@ -35,44 +36,62 @@ namespace internal {
+  inline void UpdateAllocatedBytes(int64_t diff) {
+    // max_memory_ is monotonically increasing, so we can use a relaxed load
+    // before the read-modify-write. Issue the load before everything else.
+    auto max_memory = max_memory_.load(std::memory_order_relaxed);
```
This is not required if freeing memory?

felipecrv replied:
Yes, but it doesn't get used... I will just make two functions. One for allocations and another for freeing. No need for `constexpr` but also not testing of the `diff` variable.

amoeba commented:
Is that as simple as adding `>ThreadRange(1, 32);` to the relevant benchmark's args? When I do that to the `AllocateDeallocate` benchmark (memory_pool_benchark.cc), I see about a 10x improvement at 32 threads when I comment out the entire body of `UpdateAllocatedBytes` and about half that if I just comment out what was added in Arrow 13 (https://github.com/apache/arrow/commit/ddfa8eed9b188fcc7b38767d1858c2588c588f05#diff-2111aac8ee579e238fb15d4380f2ea1e2f3e2830da939ec3f07c61dc68038d1f).

I put the full results in a Gist: https://gist.github.com/amoeba/b95102829280dbe2b1f64e6c23a5f594. A [branch on my fork](https://github.com/amoeba/arrow/commits/exp/memory-pool-bench-threads/) shows the changes I made to the benchmark and code.

felipecrv commented:
![reorderloadstores](https://github.com/apache/arrow/assets/207795/58ad547b-3374-412b-920a-8381344f7ce6)

mapleFU reviewed (approved):
Tough I still prefer relaxed. This is better than previous code. +1

zanmato1984 commented:
> > Does it actually change anything for x86?
> 
> Wouldn't acq-rel cheaper than original total-ordering?

If you mean that if acq-rel is cheaper than the original seq-cst on x86, the answer is no: RMW using acq-rel is promoted to seq-cst on x86.

If you mean that if acq-rel is cheaper than x86's "total-store-ordering" memory model, the answer is either no: x86's "total-store-ordering" guarantees acquire-release semantic (actually a little subtly stronger than acquire-release), even for plain loads and stores. Not mentioning the fact the RMW using acq-rel gets promoted to seq-cst.

As @felipecrv mentioned, only on weaker memory models such as AArch64 can acq-rel be cheaper than seq-cst - compilers can do reorderings for acq-rel that are not allowed for seq-cst. There are online discussions such as [1] implying that even solely hardware can do this too.

[1] https://stackoverflow.com/questions/65568185/for-purposes-of-ordering-is-atomic-read-modify-write-one-operation-or-two

mapleFU commented:
You're right, the generated instr would be same in x86, and weaker memory-order might able to get compiler reorder?

zanmato1984 commented:
> You're right, the generated instr would be same in x86, and weaker memory-order might able to get compiler reorder?

Yes. If the weaker memory order has the same asm as the stronger one, hardware treats them the same and only compiler can make some difference of reordering.

mapleFU commented:
Great. I'm also interested in this. I can testing on ARM CPU tomorrow

zanmato1984 commented:
I like this PR. But I have a question similar to @mapleFU 's.

IIUC, the memory pool stats update is not fully synchronized with the actual allocator action (alloc/dealloc) so it should be expected to be "accurate most of the time" rather than "accurate all the time". For example, it could be the case that two concurrent threads have the following interleaving:
1. Thread 1 allocates 1MB (actual peak memory 1MB)
2. Thread 2 allocates 2MB (actual peak memory 3MB)
3. Thread 1 updates stats (stats allocated memory 1MB, peak memory 1MB)
4. Thread 1 frees 1MB
5. Thread 1 updates stats (stats allocated memory 0MB, peak memory 1MB)
6. Thread 2 updates stats (stats allocated memory 2MB, peak memory 2MB)
...

Given the above nature of memory pool stats, should relaxed memory order be good enough (assuming it has better performance)? Though it is more "inaccurate" than acq-rel.

Thanks.

pitrou commented:
AFAIU, "relaxed" means the statistics could become entirely wrong, e.g. they could diverge more and more from the actual values. "acq-rel" ensures they converge to the actual values if no allocation happens for a long enough period of time.

mapleFU commented:
> AFAIU, "relaxed" means the statistics could become entirely wrong, e.g. they could diverge more and more from the actual values. "acq-rel" ensures they converge to the actual values if no allocation happens for a long enough period of time.

I think `compare_exchange` would gurantee the "update" maximum would be right (rather than diverge more and more from the actual values). Other operations just mean a little ( not more and more) from the actual values?

pitrou commented:
The maximum would be the least sensitive to divergence actually, become it's replacing one correct value with another correct value. Non-atomic in-place addition, though, could easily diverge if some additions are "lost".

pitrou commented:
And again, if we want the fastest possible while correct, we can invest some time porting https://travisdowns.github.io/blog/2020/07/06/concurrency-costs.html

zanmato1984 commented:
> AFAIU, "relaxed" means the statistics could become entirely wrong, e.g. they could diverge more and more from the actual values. "acq-rel" ensures they converge to the actual values if no allocation happens for a long enough period of time.

I think at least in this PR, for most of those monotonically increasing variables, relaxed atomic makes sense at certain extent:
1. Atomic fetch_add using relaxed MO doesn't lose increment, so the value will converge, eventually.
2. The "correctness" of these values depends on how they are observed. Consider a scenario where a main thread creates several concurrent worker threads and waits for them to finish:
  - If the worker threads need to see fresh enough stats numbers (assuming otherwise the program will behave unexpectedly), then acq-rel is necesary, as relaxed will cause the worker threads not able to see each other's modification from time to time (at the worst case, until the thread finished).
  - If only the main thread needs to see the stats numbers, then relaxed is good enough. Because the "happens-before" relationship established by the thread waiting will guarantee all the modifications made by the work threads visible to the main thread. As long as the increments don't get lost (and they won't, as said in 1), the numbers seen by the main thread will be correct.
  - Even for the case of "the worker threads need to see fresh enough stats numbers", though acq-rel will make the modifications immediately visible to others, the "correctness" is also arguable. Because there is no synchronization between the underlying allocator and the stats updating, as I mentioned in my previous comment.

I don't have a strong opinion on either solution, just interested in exploring all possible feasibilities. Thanks.

mapleFU commented:
Yes, I also don't have a strong opinion on either solution. I'm ok with current change

mapleFU commented:
On My MacOS M1Pro:

Before:

```
TouchArea/size:4096/real_time/threads:1                                       10.9 ns         10.9 ns     57994989 bytes_per_second=350.334G/s items_per_second=91.838M/s
TouchArea/size:4096/real_time/threads:2                                       5.60 ns         11.2 ns    124876054 bytes_per_second=681.267G/s items_per_second=178.59M/s
TouchArea/size:4096/real_time/threads:4                                       2.86 ns         11.4 ns    245521852 bytes_per_second=1.30431T/s items_per_second=350.122M/s
TouchArea/size:4096/real_time/threads:8                                       1.45 ns         11.6 ns    484530120 bytes_per_second=2.57092T/s items_per_second=690.125M/s
TouchArea/size:4096/real_time/threads:16                                      1.38 ns         12.9 ns    590918384 bytes_per_second=2.69684T/s items_per_second=723.928M/s
TouchArea/size:4096/real_time/threads:32                                      1.29 ns         12.8 ns    556564992 bytes_per_second=2.87679T/s items_per_second=772.233M/s
TouchArea/size:65536/real_time/threads:1                                       208 ns          208 ns      3317775 bytes_per_second=293.541G/s items_per_second=4.80938M/s
TouchArea/size:65536/real_time/threads:2                                       107 ns          213 ns      6526584 bytes_per_second=572.079G/s items_per_second=9.37295M/s
TouchArea/size:65536/real_time/threads:4                                      54.1 ns          216 ns     12943336 bytes_per_second=1.10126T/s items_per_second=18.4761M/s
TouchArea/size:65536/real_time/threads:8                                      27.5 ns          219 ns     25424256 bytes_per_second=2.16524T/s items_per_second=36.3268M/s
TouchArea/size:65536/real_time/threads:16                                     28.7 ns          242 ns     29298224 bytes_per_second=2.07599T/s items_per_second=34.8294M/s
TouchArea/size:65536/real_time/threads:32                                     25.1 ns          245 ns     38874976 bytes_per_second=2.37519T/s items_per_second=39.8491M/s
TouchArea/size:1048576/real_time/threads:1                                   10319 ns        10315 ns        67168 bytes_per_second=94.6367G/s items_per_second=96.908k/s
TouchArea/size:1048576/real_time/threads:2                                    5327 ns        10644 ns       133090 bytes_per_second=183.324G/s items_per_second=187.724k/s
TouchArea/size:1048576/real_time/threads:4                                    2691 ns        10758 ns       261484 bytes_per_second=362.937G/s items_per_second=371.648k/s
TouchArea/size:1048576/real_time/threads:8                                    1368 ns        10927 ns       506984 bytes_per_second=713.808G/s items_per_second=730.939k/s
TouchArea/size:1048576/real_time/threads:16                                   1218 ns        11969 ns       600560 bytes_per_second=802.048G/s items_per_second=821.297k/s
TouchArea/size:1048576/real_time/threads:32                                   1204 ns        12061 ns       557376 bytes_per_second=811.109G/s items_per_second=830.576k/s
TouchArea/size:16777216/real_time/threads:1                                 253526 ns       253370 ns         2742 bytes_per_second=61.6306G/s items_per_second=3.94436k/s
TouchArea/size:16777216/real_time/threads:2                                 139961 ns       279856 ns         4868 bytes_per_second=111.638G/s items_per_second=7.14484k/s
TouchArea/size:16777216/real_time/threads:4                                 152334 ns       609277 ns         4424 bytes_per_second=102.571G/s items_per_second=6.56452k/s
TouchArea/size:16777216/real_time/threads:8                                 153201 ns      1224218 ns         4512 bytes_per_second=101.99G/s items_per_second=6.52735k/s
TouchArea/size:16777216/real_time/threads:16                                137789 ns      1285625 ns         5296 bytes_per_second=113.398G/s items_per_second=7.25745k/s
TouchArea/size:16777216/real_time/threads:32                                126020 ns      1315402 ns         6016 bytes_per_second=123.988G/s items_per_second=7.93522k/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:1                 32.2 ns         32.1 ns     21709751 items_per_second=31.0963M/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:2                 75.9 ns          152 ns      9743782 items_per_second=13.1729M/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:4                  103 ns          413 ns      7189620 items_per_second=9.67396M/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:8                  127 ns         1017 ns      5459352 items_per_second=7.85479M/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:16                 164 ns         1637 ns      4368576 items_per_second=6.08059M/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:32                 159 ns         1610 ns      3200000 items_per_second=6.27851M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:1                34.6 ns         34.6 ns     20213807 items_per_second=28.8995M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:2                78.6 ns          157 ns      8997418 items_per_second=12.7283M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:4                98.2 ns          393 ns      6788752 items_per_second=10.1842M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:8                 130 ns         1037 ns      5417496 items_per_second=7.67272M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:16                167 ns         1659 ns      4431040 items_per_second=5.99549M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:32                161 ns         1639 ns      3200000 items_per_second=6.21225M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:1              34.5 ns         34.5 ns     20252808 items_per_second=28.9929M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:2              80.6 ns          161 ns      8622844 items_per_second=12.4024M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:4               101 ns          403 ns      7106452 items_per_second=9.91669M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:8               130 ns         1040 ns      5398432 items_per_second=7.66976M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:16              163 ns         1551 ns      4274144 items_per_second=6.12846M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:32              163 ns         1642 ns      3200000 items_per_second=6.15232M/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:1              212 ns          211 ns      3334565 items_per_second=4.71897M/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:2              743 ns         1418 ns      1107620 items_per_second=1.34635M/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:4             1094 ns         4374 ns       627532 items_per_second=914.037k/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:8             1769 ns        14099 ns       400296 items_per_second=565.15k/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:16            2024 ns        20213 ns       338096 items_per_second=494.161k/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:32            1931 ns        20111 ns       320000 items_per_second=517.831k/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:1            43.6 ns         43.6 ns     16031134 bytes_per_second=87.4222G/s items_per_second=22.9172M/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:2            82.3 ns          164 ns      7117618 bytes_per_second=46.3634G/s items_per_second=12.1539M/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:4            94.8 ns          379 ns      6668000 bytes_per_second=40.2382G/s items_per_second=10.5482M/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:8             131 ns         1044 ns      5370112 bytes_per_second=29.0769G/s items_per_second=7.62233M/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:16            163 ns         1620 ns      4179600 bytes_per_second=23.3966G/s items_per_second=6.13327M/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:32            158 ns         1597 ns      3200000 bytes_per_second=24.2022G/s items_per_second=6.34447M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:1            259 ns          259 ns      2714421 bytes_per_second=235.671G/s items_per_second=3.86124M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:2            142 ns          284 ns      4417676 bytes_per_second=430.031G/s items_per_second=7.04563M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:4            130 ns          519 ns      4000000 bytes_per_second=468.521G/s items_per_second=7.67625M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:8            131 ns         1049 ns      5247904 bytes_per_second=464.59G/s items_per_second=7.61185M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:16           156 ns         1576 ns      4479440 bytes_per_second=391.971G/s items_per_second=6.42205M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:32           154 ns         1508 ns      4565600 bytes_per_second=397.094G/s items_per_second=6.50599M/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:1        10273 ns        10270 ns        68181 bytes_per_second=95.0594G/s items_per_second=97.3409k/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:2         5309 ns        10616 ns       132128 bytes_per_second=183.938G/s items_per_second=188.353k/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:4         2739 ns        10948 ns       255152 bytes_per_second=356.54G/s items_per_second=365.097k/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:8         1511 ns        11467 ns       497368 bytes_per_second=646.375G/s items_per_second=661.888k/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:16        1220 ns        12373 ns       609424 bytes_per_second=800.314G/s items_per_second=819.522k/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:32        1223 ns        12554 ns       587936 bytes_per_second=798.588G/s items_per_second=817.754k/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:1     1325250 ns      1323670 ns          528 bytes_per_second=11.7902G/s items_per_second=754.574/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:2      689854 ns      1379529 ns          988 bytes_per_second=22.6497G/s items_per_second=1.44958k/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:4      519968 ns      2079136 ns         1320 bytes_per_second=30.0499G/s items_per_second=1.9232k/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:8      155025 ns      1238565 ns         4536 bytes_per_second=100.79G/s items_per_second=6.45058k/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:16     179778 ns      1879703 ns         3872 bytes_per_second=86.9129G/s items_per_second=5.56242k/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:32     253345 ns      2907238 ns         2048 bytes_per_second=61.6748G/s items_per_second=3.94719k/s
```

After:

```
TouchArea/size:4096/real_time/threads:1                                       10.9 ns         10.9 ns     58233346 bytes_per_second=349.422G/s items_per_second=91.5988M/s
TouchArea/size:4096/real_time/threads:2                                       5.60 ns         11.2 ns    125193588 bytes_per_second=681.472G/s items_per_second=178.644M/s
TouchArea/size:4096/real_time/threads:4                                       2.85 ns         11.4 ns    245458448 bytes_per_second=1.3062T/s items_per_second=350.63M/s
TouchArea/size:4096/real_time/threads:8                                       1.48 ns         11.8 ns    484327440 bytes_per_second=2.52042T/s items_per_second=676.57M/s
TouchArea/size:4096/real_time/threads:16                                      1.30 ns         13.0 ns    583730720 bytes_per_second=2.86103T/s items_per_second=768.003M/s
TouchArea/size:4096/real_time/threads:32                                      1.39 ns         13.0 ns    572841536 bytes_per_second=2.67056T/s items_per_second=716.874M/s
TouchArea/size:65536/real_time/threads:1                                       207 ns          206 ns      3384914 bytes_per_second=295.514G/s items_per_second=4.8417M/s
TouchArea/size:65536/real_time/threads:2                                       106 ns          212 ns      6590914 bytes_per_second=574.778G/s items_per_second=9.41716M/s
TouchArea/size:65536/real_time/threads:4                                      54.2 ns          217 ns     12962756 bytes_per_second=1126.11G/s items_per_second=18.4502M/s
TouchArea/size:65536/real_time/threads:8                                      27.4 ns          219 ns     25492952 bytes_per_second=2.17293T/s items_per_second=36.4557M/s
TouchArea/size:65536/real_time/threads:16                                     24.6 ns          247 ns     28787872 bytes_per_second=2.42679T/s items_per_second=40.7148M/s
TouchArea/size:65536/real_time/threads:32                                     24.6 ns          247 ns     33010080 bytes_per_second=2.4265T/s items_per_second=40.7098M/s
TouchArea/size:1048576/real_time/threads:1                                   10282 ns        10277 ns        67234 bytes_per_second=94.9769G/s items_per_second=97.2564k/s
TouchArea/size:1048576/real_time/threads:2                                    5324 ns        10643 ns       131616 bytes_per_second=183.436G/s items_per_second=187.838k/s
TouchArea/size:1048576/real_time/threads:4                                    2686 ns        10738 ns       255752 bytes_per_second=363.547G/s items_per_second=372.272k/s
TouchArea/size:1048576/real_time/threads:8                                    1375 ns        10961 ns       507424 bytes_per_second=710.368G/s items_per_second=727.417k/s
TouchArea/size:1048576/real_time/threads:16                                   1210 ns        12051 ns       615152 bytes_per_second=807.181G/s items_per_second=826.553k/s
TouchArea/size:1048576/real_time/threads:32                                   1190 ns        12141 ns       604992 bytes_per_second=820.868G/s items_per_second=840.569k/s
TouchArea/size:16777216/real_time/threads:1                                 273083 ns       273033 ns         2702 bytes_per_second=57.217G/s items_per_second=3.66189k/s
TouchArea/size:16777216/real_time/threads:2                                 141548 ns       283059 ns         4964 bytes_per_second=110.387G/s items_per_second=7.06475k/s
TouchArea/size:16777216/real_time/threads:4                                 158891 ns       632653 ns         4412 bytes_per_second=98.3379G/s items_per_second=6.29363k/s
TouchArea/size:16777216/real_time/threads:8                                 155452 ns      1216134 ns         4608 bytes_per_second=100.513G/s items_per_second=6.43284k/s
TouchArea/size:16777216/real_time/threads:16                                134444 ns      1315938 ns         5296 bytes_per_second=116.219G/s items_per_second=7.43803k/s
TouchArea/size:16777216/real_time/threads:32                                127480 ns      1309385 ns         5472 bytes_per_second=122.569G/s items_per_second=7.84439k/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:1                 32.1 ns         32.1 ns     21690403 items_per_second=31.1574M/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:2                 82.7 ns          165 ns      7816064 items_per_second=12.091M/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:4                  140 ns          560 ns      4934076 items_per_second=7.14323M/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:8                  173 ns         1384 ns      4104304 items_per_second=5.77505M/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:16                 286 ns         2883 ns      2458128 items_per_second=3.49514M/s
AllocateDeallocate<SystemAlloc>/size:4096/real_time/threads:32                 285 ns         2870 ns      2917728 items_per_second=3.51277M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:1                34.6 ns         34.6 ns     20263456 items_per_second=28.9233M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:2                90.9 ns          182 ns      7188236 items_per_second=11.0032M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:4                 149 ns          597 ns      4807920 items_per_second=6.69232M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:8                 171 ns         1365 ns      4155448 items_per_second=5.85232M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:16                283 ns         2822 ns      2480400 items_per_second=3.53548M/s
AllocateDeallocate<SystemAlloc>/size:65536/real_time/threads:32                283 ns         2905 ns      3027264 items_per_second=3.53694M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:1              34.5 ns         34.4 ns     20251705 items_per_second=29.0154M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:2              87.2 ns          174 ns      7039696 items_per_second=11.4744M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:4               163 ns          651 ns      5453856 items_per_second=6.14621M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:8               179 ns         1419 ns      4187808 items_per_second=5.58798M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:16              281 ns         2866 ns      2439200 items_per_second=3.56015M/s
AllocateDeallocate<SystemAlloc>/size:1048576/real_time/threads:32              283 ns         2907 ns      2820320 items_per_second=3.53071M/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:1              210 ns          210 ns      3335281 items_per_second=4.76775M/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:2              595 ns         1189 ns      1236228 items_per_second=1.68083M/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:4             1145 ns         4575 ns       663448 items_per_second=873.578k/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:8             1756 ns        13963 ns       405592 items_per_second=569.538k/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:16            2006 ns        20012 ns       351120 items_per_second=498.43k/s
AllocateDeallocate<SystemAlloc>/size:16777216/real_time/threads:32            1872 ns        19736 ns       446720 items_per_second=534.2k/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:1            43.3 ns         43.3 ns     16139653 bytes_per_second=88.1665G/s items_per_second=23.1123M/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:2            83.6 ns          167 ns      6850320 bytes_per_second=45.647G/s items_per_second=11.9661M/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:4             146 ns          583 ns      5593792 bytes_per_second=26.1403G/s items_per_second=6.85252M/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:8             194 ns         1549 ns      3623832 bytes_per_second=19.6813G/s items_per_second=5.15933M/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:16            288 ns         2903 ns      2478016 bytes_per_second=13.2574G/s items_per_second=3.47535M/s
AllocateTouchDeallocate<SystemAlloc>/size:4096/real_time/threads:32            284 ns         2891 ns      2915552 bytes_per_second=13.4396G/s items_per_second=3.52311M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:1            256 ns          256 ns      2729001 bytes_per_second=238.069G/s items_per_second=3.90053M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:2            134 ns          267 ns      5425282 bytes_per_second=456.636G/s items_per_second=7.48152M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:4            130 ns          519 ns      5629480 bytes_per_second=470.541G/s items_per_second=7.70935M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:8            207 ns         1609 ns      3285216 bytes_per_second=294.611G/s items_per_second=4.8269M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:16           262 ns         2341 ns      2897648 bytes_per_second=233.093G/s items_per_second=3.819M/s
AllocateTouchDeallocate<SystemAlloc>/size:65536/real_time/threads:32           240 ns         2212 ns      2858336 bytes_per_second=254.517G/s items_per_second=4.17001M/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:1        10957 ns        10947 ns        64417 bytes_per_second=89.1241G/s items_per_second=91.2631k/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:2         5524 ns        11038 ns       127042 bytes_per_second=176.772G/s items_per_second=181.015k/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:4         2763 ns        11041 ns       200264 bytes_per_second=353.498G/s items_per_second=361.982k/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:8         1477 ns        11775 ns       459144 bytes_per_second=661.092G/s items_per_second=676.958k/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:16        1298 ns        12362 ns       588272 bytes_per_second=752.097G/s items_per_second=770.147k/s
AllocateTouchDeallocate<SystemAlloc>/size:1048576/real_time/threads:32        1246 ns        12478 ns       583680 bytes_per_second=784.035G/s items_per_second=802.852k/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:1     1320591 ns      1320194 ns          520 bytes_per_second=11.8318G/s items_per_second=757.237/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:2      703512 ns      1406840 ns         1010 bytes_per_second=22.21G/s items_per_second=1.42144k/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:4      555389 ns      2220354 ns         1120 bytes_per_second=28.1335G/s items_per_second=1.80054k/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:8      401878 ns      3200275 ns         1744 bytes_per_second=38.88G/s items_per_second=2.48832k/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:16     158491 ns      1588349 ns         4128 bytes_per_second=98.5858G/s items_per_second=6.30949k/s
AllocateTouchDeallocate<SystemAlloc>/size:16777216/real_time/threads:32     259015 ns      2796938 ns         3232 bytes_per_second=60.3246G/s items_per_second=3.86078k/s
```

felipecrv commented:
@pitrou wrote:

> The maximum would be the least sensitive to divergence actually, become it's replacing one correct value with another correct value. At worse, it would not report the true maximum, but it would report a reasonable value nonetheless.
> 
> Non-atomic in-place addition, though, could easily diverge if some additions are "lost".

This is exactly what I concluded as well.

And I agree that the solution here would be to have multiple counters and paying the cost on loads, instead of plastering `relaxed` everywhere and getting non-sensical results.

@mapleFU @zanmato1984 I decided to have this PR focus solely on the ordering of loads and stores and will explore contention-reduction solution in another issue/PR pair. But there is also something @amoeba would like to be considered here: can we disable the stats collection altogether [1] considering its non-trivial cost?

[1] controlled by some build or runtime flag

zanmato1984 reviewed (approved):
+1

mapleFU commented:
> ![reorderloadstores](https://private-user-images.githubusercontent.com/207795/316207299-58ad547b-3374-412b-920a-8381344f7ce6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTEzODM4NzEsIm5iZiI6MTcxMTM4MzU3MSwicGF0aCI6Ii8yMDc3OTUvMzE2MjA3Mjk5LTU4YWQ1NDdiLTMzNzQtNDEyYi05MjBhLTgzODEzNDRmN2NlNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyNVQxNjE5MzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kOTIwZDAzOWU5MDU4YmQ3Zjk4MGNmZmFjNDBkNTJjNTlhYmRjMWI0ODdkOGI4MDdiZWRiMDJlMjk1NzA2YzQ4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.c8bogEryhO_qDfNNfM67aWKywedSTJrxGymS1viuk7o)

+1 Also just curious how do you draw this

felipecrv commented:
> +1 Also just curious how do you draw this

@mapleFU I paste the CSV output of the benchmarks I invoke manually, on Excel, then I make a chart based on that data.

mapleFU commented on cpp/src/arrow/memory_pool.h:
```diff
@@ -35,44 +36,68 @@ namespace internal {
+    total_allocated_bytes_.fetch_add(size, std::memory_order_acq_rel);
+    num_allocs_.fetch_add(1, std::memory_order_acq_rel);
+
+    // If other threads are updating max_memory_ concurrently we leave the loop without
```
(Actually this might make program a big slower bug making `max_memory_` more precisely?)

felipecrv replied:
I sampled with `perf` at 10000 Hertz and this was almost never hit by the sampler (I pasted the disassembly with counts and percentages on the main thread).

If threads are competing for `max_memory_`, whoever wins and pushes the update through the memory system frees all the other threads from having to update the value. Whereas before we would have every thread racing to apply their store on the value to this cache line.

mapleFU commented on cpp/src/arrow/memory_pool.h:
```diff
@@ -35,44 +36,68 @@ namespace internal {
+    // with execution. When done, max_memory and old_bytes_allocated have
+    // a higher chance of being available on CPU registers. This also has the
+    // nice side-effect of putting 3 atomic stores close to each other in the
+    // instruction stream.
```
May I also ask why here not using `std::memory_order_release` for the store instr?

felipecrv replied:
`fetch_add` is not a store, it's a load+store. The ALU needs the latest value to correctly calculate the sum, that then gets stored in the memory position.

mapleFU replied:
https://godbolt.org/z/bfKe9vG6P @felipecrv I've tested this, without geting the result, `-O2` only stores?

mapleFU replied:
https://godbolt.org/z/bfKe9vG6P Seems `fetch_add` without return calls `lock add`, and with returning calls xadd, don't know would it load

mapleFU replied:
( Maybe a sample with Release only: https://github.com/rust-lang/rust/blob/60b5ca62752ecc25d578066c8b82e1a4887267d4/library/std/src/sync/mpmc/list.rs#L241 ) I'm not sure here

felipecrv replied:
At the semantic level, there is a load happening. That's what the `acq` in `acq_rel` refers to. And at micro-architectural level, the CPU is loading the value before doing the `add` in the ALU and storing the result.

The fact that `lock add` and `lock xadd` both guarantee that load+add+store will happen atomically is an internal concern of the compiler backend when it's doing x86 instruction selection. 

The memory model of C++ the language is much richer than the actual implementation in x86. `acq_rel` (and `relaxed`, and `seq_cst`) are the technically correct [1] orders to use in read-modify-write operations even though passing `release` here would be synonym to `acq_rel` on x86.

[1] https://en.cppreference.com/w/c/atomic/memory_order

mapleFU replied:
I get your point here. `acq` is used to synchronizes-with other inc command rather than. Though personally I think we can just sync with reader? I think we can keep it here.
